pkg/vpcmodel/subnetsIPBlocks.go

```diff
@@ -0,0 +1,266 @@
+// To understand and monitor the connectivity that is potentially induced by the load balancer,
+// we create a private IP for all the load balancer subnets.
+// (we calls these private IPs fake private Ips)
+// Moreover, if a filter rule splits the subnet's cidr to few blocks, we want to create private ip per each such block.
+
+// To create a private IP which does not exist in the config, we need an unused address.
+// subnetsIPBlocks has two main purposes:
+//   1. for each subnet, it calculates splitByFiltersBlocks - the atomic blocks induced by the filters.
+//      splitByFiltersBlocks are such that:
+//   		a. the blocks are disjoint
+//	    	b. the union of the blocks is the subnet cidr
+//      we calculate splitByFiltersBlocks of a subnet by:
+//          (a) collecting all the filters' rules blocks.
+//          (b) adding allCidr block to these blocks.
+//          (c) from these blocks, create a list of disjoint blocks
+//          (d) for each of these blocks, intersect the block with the subnet cidr
+//          (e) collect all the non-empty intersections we get in (d)
+//   2. allocate a free address for the fake private IPs - to this end, we hold for each subnet freeAddressesBlocks,
+//      freeAddressesBlocks are splitByFiltersBlocks minus all the addresses that were already allocated.
+//      to get freeAddressesBlocks, we first copy splitByFiltersBlocks, then we remove the subnet already allocated addresses.
+//      (we get these addresses from the subnet reserved IP).
+//       when a free address is needed, we take the first address of the block and remove it from the free blocks list.
+
+// /////////////////////////////////////////////////////////////////////////////////////////
+// oneSubnetBlocks hold the blocks of one subnet
+type oneSubnetBlocks struct {
+	subnetOriginalBlock *ipblock.IPBlock // the block of the cidr of the subnet
+	// splitByFiltersBlocks are the atomic blocks induced by the filters, the union of this slice is the subnetOriginalBlock
+	splitByFiltersBlocks []*ipblock.IPBlock
+	// freeAddressesBlocks are the splitByFiltersBlocks minus the reserved IPs
+	// each block in splitByFiltersBlocks has a corresponding block at freeAddressesBlocks
+	freeAddressesBlocks []*ipblock.IPBlock
+	// fullyReservedBlocks is a bool per block.
+	// its true if all the addresses in the original block are reserved IP
+	// for these blocks there is no need to create private IPs
+	fullyReservedBlocks []bool
+}
+
+// SubnetsIPBlocks is a map from the subnet crn to the subnets block
+type SubnetsIPBlocks map[string]*oneSubnetBlocks
+
+// getSubnetsIPBlocks() is the main func that creates the subnetsBlocks, steps:
+// 1. get the subnets original blocks
+// 2. calculate the filters blocks
+// 3. calculate the splitByFiltersBlocks
+// 4. calculate the freeAddressesBlocks
+func GetSubnetsIPBlocks(rc *datamodel.ResourcesContainerModel, filtersCidrs []map[string][]*string,
```

Comment:
move this to pkg `ibmvpc`, since it is ibm-specific for `datamodel.ResourcesContainerModel`

Reply:
done
added https://github.com/np-guard/vpc-network-config-analyzer/issues/666